### ISSUE TITLE
`safe-super` change breaks Mixin support

### ISSUE THREAD

**Issue by Kakadus** [python/mypy#17192]
**Bug Report**

We want to indicate that a (super-)class must implement a specific protocol in order to inherit from a specific (sub-) class. We used to annotate `self` in the subclass to express this. With mypy `1.10.0` this introduces a `safe-super` error. Is there currently a way to express this requirement?

**To Reproduce**

```python
from typing import Protocol


class HasFormValid(Protocol):
    def form_valid(self, form):
        pass


class SaveValidFormMixin:

    def form_valid(self: HasFormValid, form):
        form.save()
        super().form_valid(form)

```

**Expected Behavior**

No error: Since we never actually inherit from the `HasFormValid` protocol, this should be safe.

**Actual Behavior**

mypy 1.10.0 treats `HasFormValid.form_valid` as trivial and warns about its usage. In this case, it is assured that `HasFormValid.form_valid` cannot be actually called via `super().form_valid`.

mypy <1.10.0 does not raise an error here.

**Your Environment**

- Mypy version used: mypy 1.10.0 (compiled: yes)
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12.3

**Comment by adamnovak** [comment]
I'm stuck on what might be this bug, when trying to type a class that you multiply-inherit from to wrap an implementation of an abstract method defined in an abstract base class.

If I make the wrapping mixin type inherit from the abstract base class, MyPy raises `safe-super` because I'm defining a concrete wrapper method implementation that apparently calls the trivial abstract `super()` version of the method, because I don't know how to explain to MyPy that the class must be mixed in with another class that provides the concrete implementation.

[StackOverflow](https://stackoverflow.com/questions/78377563/mypy-properly-typing-a-django-mixin-class-when-accessing-a-method-on-super/78391066#78391066) pointed to [MyPy documentation on typing mixin classes](https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes), which suggests using protocols to type `self` as shown in the example for this issue. The documentation doesn't show any more complex examples, and so doesn't help you with either this problem (where MyPy won't let you use `super()` to call a method on the protocol that you are overriding), or with the problem where, having typed `self` as the protocol, you no longer are allowed to use any of the other methods or fields of your mixin in that method (since MyPy thinks self is now the protocol, not the protocol plus the rest of the mixin).

The [MyPy documentation for `safe-super`](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-call-to-an-abstract-method-via-super-is-valid-safe-super) sort of explains the problem, but doesn't really help someone get out of a situation in which they've encountered the error. One probably doesn't *actually* mean to call the trivial abstract implementation that does nothign or raises `NotImplementedError`, so the problem is probably some kind of missing constraint (like the fact that a mixin needs to be actually mixed into the right kind of class) that isn't articulated to MyPy. The documentation doesn't provide any advice as to what that missing constraint could be or how one would go about articulating it in common situations.

I can't tell if it's really a bug in MyPy, or if there's a problem with the documentation where it doesn't provide an example of the right incantations to wrap a method in a mixin.

**Comment by adamnovak** [comment]
The way of typing mixins with protocols on `self` shown at https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes I think only allows you to promise that things exist at the level of the final mixed-in type.

You can do:
```python
from typing import Protocol


class HasFormValid(Protocol):
    def form_valid(self, form):
        pass


class SaveValidFormMixin:

    def form_valid_wrapper(self: HasFormValid, form):
        form.save()
        self.form_valid(form)
```

This works fine and isn't even a runtime recursion loop. You've told MyPy that, by the time the mixin method runs, it will have been mixed into a class that, somewhere, defines a `form_valid()`.

But to use `super().form_valid()` from anywhere in the mixin, whether the current method name is itself `form_valid` or not, it isn't enough for the method to be defined in the class inheriting the mixin. `super()` only looks later on in the method resolution order, right? So you'd somehow need to explain to MyPy that the method will only run in a class that not only defines `form_valid()`, but defines it at a lower point in the method resolution order than where the mixin will sit.

**Comment by niklasmohrin** [comment]
Yes, the problem is that we just cannot communicate that the intended way to use `SaveValidFormMixin` is to insert it into the MRO of another class in the correct order. That is, we want to say something like
```python
def form_valid(self: SuperIs[HasFormValid], form):
    return super().form_valid(form)
    #      ^^^^^^^ inferred to be HasFormValid
```
but there is no way to do this.

Still, the current warning

> Call to abstract method "form_valid" of "HasFormValid" with trivial body via super() is unsafe

is just wrong / misleading, because the call to `form_valid` can be (i) non-trivial if the mixin is used as intended, or (ii) an attribute error because `HasFormValid` is a protocol and thus doesn't need to be in the MRO at runtime.
